### ISSUE TITLE
spec: inconsistency about converting numeric constants to strings

### ISSUE THREAD

**Issue by mdempsky** [golang/go#24745]
In section "Conversions", the Go spec says:

> A constant value x can be converted to type T if x is representable by a value of T. As a special case, an integer constant x can be converted to a string type using the same rule as for non-constant x.

Note that "integer constants" are distinct from "rune constants", "float constants", etc.

However, the spec also includes two examples of conversions of rune literals to type string: `string('x')` and `string('a')`.

cmd/compile and go/types allow these conversions. I assume gccgo does too, as utf8.RuneError is an untyped rune constant, and `string(utf8.RuneError)` appears a few times in the standard library.

I think the wording should be tweaked to "an integer **or rune** constant x".

Aside: I'd think we should relax it further to allow any integer *value*. We allow `make([]byte, 1.0)`, so it seems inconsistent to not allow `string(1.0)`.

/cc @griesemer

**Comment by mdempsky** [comment]
Somewhat related: #21982.

**Comment by mdempsky** [comment]
> If the left operand of a constant shift expression is an untyped constant, the result is an integer constant;

This suggests in the code below that `'a' << 1` should be an integer constant, so `x` should be given the default type `int`, and the blank assignment should succeed.

	var x = 'a' << 1
	var _ int = x

However, cmd/compile and go/types both give `x` the type `rune`, so the assignment is rejected.

**Comment by go101** [comment]
doesn't integer constants include rune constants?

**Comment by cznic** [comment]
I also think 'x' _is_ an integer constant. It's just its default type is `rune`.

**Comment by mdempsky** [comment]
`rune` is an alias for `int32`, which is defined as an integer type. But there doesn't appear to be any wording to suggest that "rune constants" are a subclass of "integer constants."

**Comment by cznic** [comment]
I think that `'x'` is not (yet) a `rune` constant, but an untyped integer constant. In certain contexts, like in `foo := 'x'` it _becomes_ a `rune`. In `string('x')` its default type (`rune`) is IMO not relevant. I think that it's in the later case actually (due to context) first converted to `int`, not `rune`, like in the (IMHO) equivalent `string(120)`.

**Comment by mdempsky** [comment]
> I think that 'x' is not (yet) a rune constant, but an untyped integer constant.

That's not what the Go spec says:

> A rune literal represents a **rune constant**, an integer value identifying a Unicode code point.

Rune constants are integer values, but not integer constants.

**Comment by griesemer** [comment]
I agree with @mdempsky that there is at least some imprecision if not inconsistency in the spec in this area. We use mildly similar terminology for different purposes (integer value vs integer constant vs integer vs int type, etc.) and these terms are vaguely defined if at all.

I think we need to clearly define these terms in one place (consistent with existing use) and then this issue (and related ones) should become clear.
